### PR TITLE
Implement build logs

### PR DESCRIPTION
**Pull request by TheRealJon**
# Description
Add logs to Build detail view.

# Changes
- Create new ResourceLog component which accepts props to make API request for log and manages log status.
    - Broke out a few stateless components like LogControls and ContainerDropdown which are used to render the the controls above the log window.
    - Added a new log status, 'eof', which will disable the streaming controls and stop requests to the API
    - Accepts an `eof` property from the parent which is used to trigger the 'eof' state.
    - Log status changes are now all funneled through `_updateLogStatus` function to always check for eof condition before updating to a different state.
    
- Create BuildLog component which uses the new ResourceLog component
- Refactor PodLogs component to use ResourceLog
- Add handling for case where Build/Container is no longer running and streaming controls should be disabled

# Screen Shots
Pod Logs (without play/pause button)
screenshot-localhost-9000-2018 05 24-11-29-39

Build Logs (no dropdown or label since it is in the page heading)
screenshot-localhost-9000-2018 05 24-11-28-42

### CODE REVIEW

**Comment by TheRealJon** [comment]
@ggreer @kans and/or @amrutac,  could you give me some feedback on this one. It's definitely still a work in progress. As you can see in the TODO list, there is still a bug I've been trying to figure out. The first time visiting a log view in either pods or builds, it is initially stuck in the 'loading' state and I can't quite figure out why. On a page reload or forcing a state change by scrolling the log, it updates, and everything works as expected, but I'm having trouble pinpointing whats causing the state not to be set correctly at first. The log state should be set to 'streaming' when the '_loadStart' callback is triggered by the XMLHttpRequest 'loadstart' event, but that doesn't seem to happen.

File: frontend/public/components/_toggle-play.scss [hunk 1]
```diff
@@ -6,6 +6,13 @@
   width: 32px;
   padding: 0;
 
+  &:disabled {
+    border-color: $color-grey-border !important;
+    &:before {
+      content: '\f05e' !important;
```
Comment (rhamilto) [inline frontend/public/components/_toggle-play.scss]: Here and on line 17 it'd be best to use the SASS var. https://github.com/FortAwesome/Font-Awesome/blob/v4.7.0/scss/_variables.scss#L66

File: frontend/public/components/utils/resource-log.jsx [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+// Component to render resource dropdown or just the resource name if only
+// one resource is provided.
+const ContainerSelect = (props) => {
+  const currentResourceNameWithIcon = <NameWithIcon name={props.currentResource} kind={props.kind} />;
+
+  if (props.selectableResources.length > 1) {
+    const dropdownItems = _.map(props.selectableResources, (name) => {
+      return <NameWithIcon name={name} kind={props.kind} />;
+    });
+    return <Dropdown className="btn-group" items={dropdownItems} title={currentResourceNameWithIcon} onChange={this._selectResource} />;
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx]: Missed `this._selectResource`, should be `props.selectResource`

File: frontend/public/components/utils/resource-log.jsx [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  }
+
+  componentDidMount() {
+    super.componentDidMount();
+    this._beginStreaming();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // If container changed, restart streaming
+    if (nextProps.container !== this.props.container) {
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx]: Added to handle container changes.

File: frontend/public/components/utils/resource-log.jsx [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    const namespace = _.get(props.obj, 'metadata.namespace');
+    const objName = _.get(props.obj, 'metadata.name');
+
+    // If the container changed, reset log state to loading, otherwise keep it as is
+    const initialLogState = this.props.container !== props.container ? 'loading' : this.state.logState;
+
+    // If log is finished, change to eof log state
+    newState.logState = props.eof ? 'eof' : initialLogState;
+
+    newState.logURL = this._logURL(props.model, namespace, objName, props.currentResource);
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx]: Instead of passing the entire props object to the _logUrl function, I parse out the arguments that are needed and pass them in.

File: frontend/public/components/utils/resource-log.jsx [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+      queryParams: {
+        container: container,
+        follow: 'true',
+        tailLines: this._buffer.maxSize
+      }
+    });
+  }
+
+  // Updates log state as long as eof has not been reached
+  _updateLogState(newState) {
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx]: Only update `logState` if `eof` property is false. Prevents changing the logState when it is no longer streaming.

File: frontend/public/components/utils/resource-log.jsx [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    // If log is at eof, don't update state to anything other than eof
+    this.setState({
+      logState: this.props.eof ? 'eof' : newState,
+    });
+  }
+
+  // use _updateLogState toggle streaming/paused state instead of directly
+  // setting logState
+  _toggleStreaming() {
+    this._updateLogState(this.state.logState === 'streaming' ? 'paused' : 'streaming');
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx]: All updates to logState go through the _updateLogState function.

File: frontend/public/components/utils/resource-log.jsx [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        this._resetPendingReload();
+        return;
+      }
+      this._beginStreaming();
+    }, wait);
+  }
+
+  // Callback which handles the XMLHttpRequest "loadstart" event
+  _loadStarted() {
+    this._updateLogState('streaming');
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx]: All updates to `logState` go through `_updateLogState` function

File: frontend/public/components/utils/resource-log.jsx [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    clearTimeout(this._pendingReload);
+    this._pendingReload = null;
+    this._buffer = lineBuffer(1000);
+
+    this._stream = stream(this.state.logURL, this._loadStarted.bind(this), this._processData.bind(this));
+    this._stream.promise
+      .then(() => {
+
+        // Resource is no longer running/generating new log content, so stop streaming
+        if (this.props.eof){
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx]: Stop streaming if `eof` property is true. This means the container or build is no longer running.

File: frontend/public/components/utils/resource-log.jsx [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        this._touchLoadTime();
+        if (!this._pendingReload) {
+          this._resetPendingReload();
+        }
+      });
+  }
+
+  render() {
+    return <div className="co-m-pane__body">
+      <LogControls {...this.props}
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx]: Use LogControls component instead of doing a bunch of logic to render the elements inline here.

File: frontend/public/components/utils/resource-log.jsx [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+        logState={this.state.logState}
+        toggleStreaming={this._toggleStreaming} />
+      <LogWindow buffer={this._buffer}
+        logName={this.props.currentResource}
+        logState={this.state.logState}
+        updateLogState={this._updateLogState}
+        loadGeneration={this.state.loadTime}
+        eof={this.props.eof} />
```
Comment (TheRealJon) [inline frontend/public/components/utils/resource-log.jsx]: `eof` prop is used in LogWindow to disable log state changes on scroll.

File: frontend/public/components/utils/resource-log.jsx [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    this._endStreaming();
+  }
+
+  _initialState(props = this.props) {
+    const newState = {};
+    const namespace = _.get(props.obj, 'metadata.namespace');
+    const objName = _.get(props.obj, 'metadata.name');
+
+    // If the container changed, reset log state to loading, otherwise keep it as is
+    const initialLogState = this.props.container !== props.container ? 'loading' : this.state.logState;
```
Comment (amrutac) [inline frontend/public/components/utils/resource-log.jsx]: > there is still a bug I've been trying to figure out. The first time visiting a log view in either pods or builds, it is initially stuck in the 'loading' state and I can't quite figure out why. On a page reload or forcing a state change by scrolling the log, it updates, and everything works as expected

This is probably causing loading error you mentioned, did you mean to check <code>this.props. currentResource</code> in here?

Comment (TheRealJon) [reply frontend/public/components/utils/resource-log.jsx]: I think this function is going to go away. I'm refactoring now to use`getDerivedStatFromProps` so that this function won't be necessary.

Comment (TheRealJon) [reply frontend/public/components/utils/resource-log.jsx]: And yes, I also caught the missed `container` reference yesterday. It will be fixed in my next commit.

**Comment by TheRealJon** [comment]
For now, I think I am going to move forward to get this working with the XMLHttpRequest, but I think the log feature would make a lot more sense using a WebSocket. I might work out the kinks in this PR and let it merge, then work on a follow up PR to switch over to use WebSockets.

File: frontend/public/components/build-logs.jsx [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+export class BuildLogs extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      eof: false
+    };
+  }
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    const eof = !(_.get(nextProps.obj, 'status.phase') === 'Running');
```
Comment (kans) [inline frontend/public/components/build-logs.jsx]: Make this a `!==`

Comment (TheRealJon) [reply frontend/public/components/build-logs.jsx]: 🤦‍♂️ Can't believe I did that.

**Comment by TheRealJon** [comment]
@openshift/team-ux-review Sorry, forgot to tag this. I added a new state to the log window which disables the play/pause button at the top and displays "Log stream has ended from" before the container/build name.

File: frontend/public/components/pod-logs.jsx [hunk 13]
```diff
@@ -1,194 +1,78 @@
...
+    if (_.isEmpty(prevState.currentContainer) && newState.containers.length > 0) {
+      newState.currentContainer = newState.containers[0];
     }
 
-    const currentContainer = newState.currentContainer || this.state.currentContainer;
-    newState.logURL = this._logURL(obj, currentContainer);
-
+    const currentContainer = newState.currentContainer || prevState.currentContainer;
+    newState.eof = isContainerTerminated(currentContainer);
     return newState;
```
Comment (kans) [inline frontend/public/components/pod-logs.jsx]: Typically, we'd only want to do this if `newState` differs from the old state, but I doubt Pods are updated often enough to impact performance.

File: frontend/public/components/pod-logs.jsx [hunk 14]
```diff
@@ -1,194 +1,78 @@
...
-    this._toggleLogState = this._toggleLogState.bind(this);
 
     this.state = {
-      containerNames: [],
-      currentContainer: '',
-      logURL: '',
-      loadTime: 0,
-      logState: 'loading'
+      containers: [],
+      currentContainer: {},
```
Comment (kans) [inline frontend/public/components/pod-logs.jsx]: It looks like we only ever reference `currentContainer.name` - can we just track that instead?

File: frontend/public/components/utils/resource-log.jsx [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    }
+  }
+
+  componentWillUnmount() {
+    super.componentWillUnmount();
+    this._endStreaming();
+  }
+
+  // Updates log state as long as eof has not been reached
+  _updateStatus(newState) {
```
Comment (kans) [inline frontend/public/components/utils/resource-log.jsx]: This variable is really `newStatus`

File: frontend/public/components/utils/resource-log.jsx [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    const newState = {};
+
+    if(nextProps.eof){
+      newState.status = prevState.status === 'paused' ? 'paused' : 'eof';
+    }
+
+    newState.url = resourceURL(modelFor(nextProps.kind), {
```
Comment (kans) [inline frontend/public/components/utils/resource-log.jsx]: Would you move this calculation to _beginStreaming since thats the only thing that uses it (`url` is not touched in render). I realize you inherited it, but `url` never belonged in state to begin with.

File: frontend/public/components/pod-logs.jsx (hunk 14 above)
Comment (TheRealJon) [reply frontend/public/components/pod-logs.jsx]: On line 52 and 58 I'm using currentContainer.status to check if the container is terminated.

**Comment by beanh66** [comment]
Thanks for chatting @TheRealJon, we'll post some design updates asap!

**Comment by spadgett** [comment]
/hold
(needs to be on its own line)

Suggest we make sure the container dropdown is consistent between the logs tab and terminal tab

**Comment by TheRealJon** [comment]
> Suggest we make sure the container dropdown is consistent between the logs tab and terminal tab

@spadgett I don't see a terminal tab in Tectonic console. Is this a feature that's in progress?

**Comment by spadgett** [comment]
It's on the pod page. You might need to rebase since it's pretty new (#34)

**Comment by beanh66** [comment]
@TheRealJon A few changes to suggest: 

1. Can we keep the container dropdown selector there regardless of the number of option inside because even with one option, it provides context and the spacing seems off without it.  
2. When the pause/play button isn't available because the build is complete/ logs have ended, we would like to hide the button altogether and just push the text to the left. 
3. In general can you check on the spacing and alignment? Seems like we can reduce the space before the container dropdown.
4. OpenShift currently has a save action on the logs page, so can we bring that over and use the Tectonic [Download] button style presented on the YAML page? 

See below: 
log

FYI @cshinn @serenamarie125 @tlwu2013

**Comment by TheRealJon** [comment]
>  1. Can we keep the container dropdown selector there regardless of the number of option inside because even with one option, it provides context and the spacing seems off without it.
> 2. When the pause/play button isn't available because the build is complete/ logs have ended, we would like to hide the button altogether and just push the text to the left.
> 3. In general can you check on the spacing and alignment? Seems like we can reduce the space before the container dropdown.
> 4. OpenShift currently has a save action on the logs page, so can we bring that over and use the Tectonic [Download] button style presented on the YAML page? 

@beanh66 I'll start working on 1-3. 
I have a separate sub-task for 4: https://jira.coreos.com/browse/CONSOLE-403. Could you put any design comments for that in the JIRA issue so that I can reference them there when I start implementing that feature?

**Comment by TheRealJon** [comment]
@beanh66 @cshinn @serenamarie125 @tlwu2013 Also, do we want there to be a dropdown for builds? I don't think the use case exists for multiple builds in the build log view. I might be wrong, but it seems like by the time you reach the build log view, you have already drilled down to a build detail page, which means the context of the log is that of one specific build object.

**Comment by spadgett** [comment]
I'd vote for taking the dropdown off the build log page since it's always going to be the build name. It just repeats the header above.

I agree it makes sense to always have the container dropdown on the pod page.

**Comment by beanh66** [comment]
@TheRealJon Oh got it. Sorry I think we had in mind that there could be options in the dropdown. If there would never be other options, I agree with you and @spadgett on removing the dropdown. With that said, can we fix the spacing and alignment so the object looks more part of the sentence?

**Comment by TheRealJon** [comment]
@spadgett @kans @chart I think I have addressed all concerns, please let me know if you have any other feedback. Otherwise, I think this is ready to squash and merge.

**Comment by TheRealJon** [comment]
For those reviewing this PR, I just squashed and removed the hold. My last couple of commits addressed a few design issues, and screenshots of those changes are in the description. I think this is ready to merge.

**Comment by beanh66** [comment]
Looks great, thanks @TheRealJon

**Comment by TheRealJon** [comment]
/hold
Just found a bug.
1. View log for an existing build
2. Select 'rebuild' from 'Actions' menu
3. Return to the log tab

Expected Result: Log status should be 'streaming', play/pause button should be shown, and the log status message should say 'Log is streaming.'
Actual Result: Log status is still 'eof', play/pause button is abscent,  and log status message reads 'Log stream has ended.'

**Comment by spadgett** [comment]
Rebuild creates a new build. (Arguably the action isn't well named.) Are you looking at the log for the new build?

**Comment by TheRealJon** [comment]
Yes, the log is streaming the contents from the new build, but the status at the top is not updating accordingly. When you select rebuild, you are automatically redirected to the build details page for the new build, then when you navigate to the 'Logs' tab, for some reason the log state is not updating. I inspected the props for the ResourcLog component, and the `eof` prop is updating, but that is not propagating to the `status` state of the ResourceLog.

**Comment by TheRealJon** [comment]
/hold cancel

Fixed bug. The ResourceLog component will not update to a status other than 'eof' once that condition is reached. The problem was that I was detecting any Build phase other than 'Running' as an eof condition. Build phases 'New' and 'Pending' should not be considered eof. Now only 'Cancelled', 'Complete', 'Error', and 'Failed' phases are considered eof.

File: frontend/public/components/pod-logs.jsx [hunk 17]
```diff
@@ -1,194 +1,79 @@
...
-      queryParams: {
-        container: currentContainer,
-        follow: 'true',
-        tailLines: this._buffer.maxSize
-      }
+    const containers = _.get(nextProps.obj, 'status.containerStatuses', []);
+    newState.containers = _.map(containers, (container) => {
+      return {
+        name: container.name,
+        status: container.state
```
Comment (kans) [inline frontend/public/components/pod-logs.jsx]: The only usage of `status` is in `containerIsRunning`.  It looks like that value is always turned into `!eof`.  Why not store the actual value of `eof`?

File: frontend/public/components/pod-logs.jsx [hunk 18]
```diff
@@ -1,194 +1,79 @@
...
+    if (_.isEmpty(prevState.currentContainer) && newState.containers.length > 0) {
+      newState.currentContainer = newState.containers[0];
+    }
 
-  _loadStarted() {
-    this.setState({
-      logState: 'streaming'
-    });
-  }
+    const currentContainer = newState.currentContainer || prevState.currentContainer;
```
Comment (kans) [inline frontend/public/components/pod-logs.jsx]: This code is hard to follow:
```javascript
  if (_.isEmpty(prevState.currentContainer) && newState.containers.length > 0) {
      newState.currentContainer = newState.containers[0];
 }

const currentContainer = newState.currentContainer || prevState.currentContainer;
```

I think you can simplify it to the following (if you store `eof` in the containers instead of `status`):

```javascript
newState.currentContainer = prevState.currentContainer || newState.containers[0];
```

File: frontend/public/components/pod-logs.jsx [hunk 19]
```diff
@@ -1,194 +1,79 @@
...
-        this._loadStarted();
-        this._buffer.push(`Error: ${why}`);
-        this._touchLoadTime();
-        if (!this._pendingReload) {
-          this._resetPendingReload();
-        }
-      });
+  _selectContainer(index) {
+    const currentContainer = this.state.containers[index];
+    const eof = !containerIsRunning(currentContainer);
```
Comment (kans) [inline frontend/public/components/pod-logs.jsx]: It is unexpected that selecting a container would have this side effect because `eof` is a property of the container.  If you change `status` above, you can clean up this code too.

**Comment by TheRealJon** [comment]
@kans I think I addressed all of your feedback. Went ahead and squashed the last commit, so it should be ready whenever you are.

**Comment by ggreer** [comment]
jenkins rebuild

**Comment by ggreer** [comment]
jenkins rebuild

**Comment by ggreer** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image has been pushed to Quay with tag `pr_30_build_300`
